operator/pkg/utils/crypto.go

```diff
@@ -0,0 +1,152 @@
+package utils
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"fmt"
+	"github.com/pavel-v-chernykh/keystore-go"
+	corev1 "k8s.io/api/core/v1"
+	"math/big"
+	"time"
+)
+
+
+func setupKey() (*big.Int, time.Time, *rsa.PrivateKey, string, time.Time, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	var serialNumber *big.Int
+	buffer := bytes.NewBufferString("")
+	var privBytes []byte
+	var err error
+	if serialNumber, err = rand.Int(rand.Reader, serialNumberLimit); err == nil {
+		notBefore := time.Now()
+		var priv *rsa.PrivateKey
+		if priv, err = rsa.GenerateKey(rand.Reader, 4096); err == nil {
+			notAfter := notBefore.Add(365 * 24 * time.Hour)
+			if privBytes, err = x509.MarshalPKCS8PrivateKey(priv); err == nil {
+				if err = pem.Encode(buffer, &pem.Block{Type: "PRIVATE KEY", Bytes: privBytes}); err == nil {
+					return serialNumber, notBefore, priv, buffer.String(), notAfter, err
+
+				}
+			}
+		}
+	}
+
+	return nil, time.Time{}, nil, "", time.Time{}, err
+}
+
+func GetNewCAandKey(leafdomain, namespace string) (keypem, certpem string, err error) {
+	serialNumber, notBefore, priv, privPem, notAfter, err := setupKey()
+	if err == nil {
+		buffer := bytes.NewBufferString("")
+		template := x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				CommonName:   fmt.Sprintf("%s.%s.svc", leafdomain, namespace),
+				Organization: []string{"Cassandra Kubernetes Operator By Datastax"},
+			},
+			NotBefore: notBefore,
+			NotAfter:  notAfter,
+
+			IsCA:                  true,
+			KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			BasicConstraintsValid: true,
+			DNSNames:              []string{fmt.Sprintf("cassandradatacenter-webhook-service.%s.svc", namespace)},
+		}
+		var derBytes []byte
+		if derBytes, err = x509.CreateCertificate(rand.Reader, &template, &template, &priv.PublicKey, priv); err == nil {
+			if err = pem.Encode(buffer, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}); err == nil {
+				cert := buffer.String()
+				return privPem, cert, nil
+			}
+		}
+	}
+	return "", "", err
+}
+
+func prepare_ca(ca *corev1.Secret) (ca_cert_bytes []byte, ca_certificate *x509.Certificate, ca_key *rsa.PrivateKey, err error) {
+	ca_certificate_pem, _ := pem.Decode(ca.Data["cert"])
+	ca_cert_bytes = ca_certificate_pem.Bytes
+	ca_key_block, _ := pem.Decode(ca.Data["key"])
+	if untyped_ca_key, ca_key_err := x509.ParsePKCS8PrivateKey(ca_key_block.Bytes); ca_key_err != nil {
+		err = ca_key_err
```

Comment:
err is not being passed in by reference.  I don't think the value of err in the caller will be effected unless you pass "err *error" as the parameter.  In any case, why aren't we just returning error instead of modifying the parameter?

Reply:
Err is not being passed in at all. Err is a [named return](https://tour.golang.org/basics/7)

Reply:
oh, ok.  TIL